src/main/java/com/dataloom/authorization/HazelcastAuthorizationService.java

```diff
@@ -24,25 +24,35 @@
 import com.google.common.collect.ImmutableSet;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
+import com.hazelcast.durableexecutor.DurableExecutorService;
 import com.kryptnostic.datastore.util.Util;
 
 public class HazelcastAuthorizationService implements AuthorizationManager {
     private static final Logger logger = LoggerFactory.getLogger( AuthorizationManager.class );
 
     private final IMap<AceKey, Set<Permission>> aces;
     private final AuthorizationQueryService     aqs;
+    private final DurableExecutorService        executor;
 
     public HazelcastAuthorizationService( HazelcastInstance hazelcastInstance, AuthorizationQueryService aqs ) {
         aces = hazelcastInstance.getMap( HazelcastMap.PERMISSIONS.name() );
         this.aqs = checkNotNull( aqs );
+        this.executor = hazelcastInstance.getDurableExecutorService( "default" );
     }
-
+    
+    private void submitNewPermissionsToElasticsearch( List<AclKeyPathFragment> key, Principal principal ) {
+    	Set<Permission> permissions = aces.get( new AceKey( key, principal ) );
+    	// executor.submit ... updateEntitySetAclsInElasticsearch( entitySetId, principal, permissions)
+    	// entitySet.permissions.<principal.type>
+    }
+    
     @Override
     public void addPermission(
             List<AclKeyPathFragment> key,
             Principal principal,
             Set<Permission> permissions ) {
         aces.executeOnKey( new AceKey( key, principal ), new PermissionMerger( permissions ) );
+        submitNewPermissionsToElasticsearch( key, principal );
```

Comment:
I would remove all these one off calls to submit new permissions here and instead use a latch, hooking the controller update call to post an AclData object the event bus, and then have a ElastichSearchUpdateService that listens for events.